File: backend/src/tasks/was_sync.ts
import { CommandOptions } from './ecs-client';
import {
  connectToDatabase,
  connectToDatalake2,
  Organization,
  DL_Organization,
  WasFinding
} from '../models';
import { plainToClass } from 'class-transformer';
import axios from 'axios';
import saveWasFindingToDb from './helpers/saveWasFindingToDb';

interface WasFindingResult {
  finding_uid: string;
  webapp_id: string;
  was_org_id: string;
  owasp_category: string;
  severity: number;
  times_detected: number;
  base_score: number;
  temporal_score: number;
  fstatus: string;
  last_detected: Date;
  first_detected: Date;
  is_remediated: boolean;
  potential: boolean;
  webapp_url: string;
  webapp_name: string;
  name: string;
  cvss_v3_attack_vector: string;
  cwe_list: number[];
  wasc_list: any[];
  last_tested: Date;
  fixed_date: Date;
  is_ignored: boolean;
  url: string;
  qid: number;
}
interface TaskResponse {
  task_id: string;
  status: 'Processing' | 'Pending' | 'Completed' | 'Failure';
  result?: {
    total_pages: number;
    current_page: number;
    data: WasFindingResult[];
  };
  error?: string;
}

const sleep = (milliseconds) => {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
};

const fetchWasFindingTask = async (org_acronym: string, page: number) => {
  console.log('Creating task to fetch WAS Finding data');
  try {
    console.log(String(process.env.CF_API_KEY));
    const response = await axios({
      url: 'https://api.staging-cd.crossfeed.cyber.dhs.gov/pe/apiv1/was_findings',
      method: 'POST',
      headers: {
        Authorization: String(process.env.CF_API_KEY),
        access_token: String(process.env.PE_API_KEY),
        'Content-Type': '' // This is needed or else it breaks because axios defaults to application/json
      },
      data: {
        org_acronym: org_acronym,
        page: page,
        per_page: 100 // Tested with 150 and 200 but this results in 502 errors on certain pages with a lot of CPEs
      }
    });
    if (response.status >= 200 && response.status < 300) {
      console.log('Task request was successful');
    } else {
      console.log('Task request failed');
    }
    return response.data as TaskResponse;
  } catch (error) {
    console.log(`Error making POST request: ${error}`);
  }
};

const fetchWasFindingData = async (task_id: string) => {
  console.log('Creating task to fetch was_findings data');
  try {
    const response = await axios({
      url: `https://api.staging-cd.crossfeed.cyber.dhs.gov/pe/apiv1/was_findings/task/${task_id}`,
      method: 'GET',
      headers: {
        Authorization: String(process.env.CF_API_KEY),
        access_token: String(process.env.PE_API_KEY),
        'Content-Type': '' // This is needed or else it breaks because axios defaults to application/json
      }
    });
    if (response.status >= 200 && response.status < 300) {
      console.log('Request was successful');
    } else {
      console.log('Request failed');
    }
    return response.data as TaskResponse;
  } catch (error) {
    console.log(`Error making GET request: ${error}`);
  }
};

export const handler = async (commandOptions: CommandOptions) => {
  console.log(
    `Scanning PE database for vulnerabilities & services for all orgs`
  );
  try {
    // Get all organizations
    await connectToDatabase();
    const allOrgs: Organization[] = await Organization.find();

    // For each organization, fetch Was finding data
    for (const org of allOrgs) {
      console.log(
        `Scanning PE database for vulnerabilities & services for ${org.acronym}, ${org.name}`
      );
      let datalakeConnnection = await connectToDatalake2();
      let dl_org = datalakeConnnection.getRepository(DL_Organization);
      const organization = await dl_org.findOne({
        where: { acronym: org.acronym }
      });
      console.log(`queried org ${organization}`);

      if (!organization) {
        console.log(
          `No organization found in the datalake with the acronym '${org.acronym}'`
        );
        continue;
      }
      let done = false;
      let page = 1;
      let total_pages = 1;

      while (!done) {
        let finished = false;

        console.log(`Fetching page ${page} of page ${total_pages}`);
        let taskRequest = await fetchWasFindingTask(org.acronym, page);
        console.log(`Fetching `);
        if (taskRequest?.task_id) {
          while (!finished) {
            if (
              taskRequest?.status === 'Processing' ||
              taskRequest?.status === 'Pending'
            ) {
              await new Promise((r) => setTimeout(r, 1000));
              taskRequest = await fetchWasFindingData(taskRequest.task_id);
            } else if (taskRequest?.status === 'Completed') {
              for (const finding of taskRequest?.result?.data ?? []) {
                try {
                  const finding_obj: WasFinding = plainToClass(WasFinding, {
                    id: finding.finding_uid,
                    webappId: finding.webapp_id,
                    wasOrgId: finding.was_org_id,
                    owaspCategory: finding.owasp_category,
                    severity: finding.severity,
                    timesDetected: finding.times_detected,
                    baseScore: finding.base_score,
                    temporalScore: finding.temporal_score,
                    fstatus: finding.fstatus,
                    lastDetected: finding.last_detected,
                    firstDetected: finding.first_detected,
                    isRemediated: finding.is_remediated,
                    potential: finding.potential,
                    webappUrl: finding.webapp_url,
                    webappName: finding.webapp_name,
                    name: finding.name,
                    cvssV3AttackVector: finding.cvss_v3_attack_vector,
                    cweList: finding.cwe_list,
                    wascList: finding.wasc_list,
                    lastTested: finding.last_tested,
                    fixedDate: finding.fixed_date,
                    isIgnored: finding.is_ignored,
                    url: finding.url,
                    qid: finding.qid,
                    organization: { id: organization.id }
                  });

                  await saveWasFindingToDb(finding_obj);
                } catch (e) {
                  console.error('Could not save WAS finding. Continuing.');
                  console.error(e);
                }
              }
              total_pages = taskRequest?.result?.total_pages || 1;
              const current_page = taskRequest?.result?.current_page || 1;
              if (current_page >= total_pages) {
                done = true;
                console.log(`Finished fetching WAS data for ${org.acronym}`);
              }
              finished = true;
              console.log(`Task completed successfully for page: ${page}`);
              page = page + 1;
            } else {
              finished = true;
              console.error(
                `Failed to recieve WAS finding task for org: ${org.acronym} on page ${page}`
              );
              page = page + 1;
            }
          }
        } else {
          done = true;
          console.log(
            `Error fetching WAS data: ${taskRequest?.error} and status: ${taskRequest?.status}`
          );
        }
      }
    }
  } catch (e) {
    console.error('Unknown failure.');
    console.error(e);
  }
  await sleep(1000);
  console.log(`Finished retrieving WAS findings for all orgs`);
};

File: backend/src/xfd_django/xfd_api/tasks/was_sync.py
"""WasSync scan."""
# Standard Python Libraries
import datetime
import logging
import os
import time
from typing import Optional

# Third-Party Libraries
import django
from django.conf import settings
import requests
from xfd_api.helpers.date_time_helpers import calculate_days_back
from xfd_mini_dl.models import Organization, WasFindings

# Django setup
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "xfd_django.settings")
os.environ["DJANGO_ALLOW_ASYNC_UNSAFE"] = "true"
django.setup()

LOGGER = logging.getLogger(__name__)

# Constants
MAX_RETRIES = 3  # Max retries for failed tasks
TIMEOUT = 60  # Timeout in seconds for waiting on task completion

headers = settings.DMZ_API_HEADER


def handler(event):
    """Retrieve and save WAS Findings from the DMZ."""
    try:
        main()
        return {
            "status_code": 200,
            "body": "DMZ WAS Finding sync completed successfully.",
        }
    except Exception as e:
        return {"status_code": 500, "body": str(e)}


def main():
    """Fetch and save DMZ WAS Findings."""
    try:
        all_orgs = Organization.objects.all()
        # all_orgs = Organization.objects.filter(acronym__in=['USAGM', 'DHS'])

        since_timestamp_str = calculate_days_back(5)

        for org in all_orgs:
            LOGGER.info("Processing organization: %s, %s", org.acronym, org.name)
            done = False
            page = 1
            total_pages = 2
            per_page = 200
            retry_count = 0

            while not done:
                data = fetch_dmz_was_findings_task(
                    org.acronym, page, per_page, since_timestamp_str
                )
                if not data or data.get("status") != "Processing":
                    LOGGER.warning(
                        "Failed to start Was Finding sync task for org: %s, %s",
                        org.acronym,
                        org.name,
                    )

                    retry_count += 1

                    if retry_count >= MAX_RETRIES:
                        LOGGER.warning(
                            "Max retries reached for org: %s. Moving to next organization.",
                            org.acronym,
                        )
                        break  # Skip to next organization

                    time.sleep(5)
                    continue

                response = fetch_dmz_was_finding_data(data.get("task_id", None))
                while response and response.get("status") == "Pending":
                    time.sleep(1)
                    response = fetch_dmz_was_finding_data(data.get("task_id", None))

                if response and response.get("status") == "Completed":
                    was_finding_array = response.get("result", {}).get("data", [])
                    total_pages = response.get("result", {}).get("total_pages", 1)
                    current_page = response.get("result", {}).get("current_page", 1)
                    LOGGER.info("findings: %s", was_finding_array)
                    save_findings_to_db(was_finding_array, org)

                    if current_page >= total_pages:
                        done = True
                    page += 1
                else:
                    raise Exception(
                        "Task error: {error} - Status: {status}".format(
                            error=response.get("error"), status=response.get("status")
                        )
                    )
    except Exception as e:
        LOGGER.error("Failed to in main: %s", e)


def fetch_dmz_was_findings_task(org_acronym, page, per_page, since_timestamp):
    """Fetch Was Finding task id."""
    LOGGER.info("Fetching WAS finding task for organization: %s", org_acronym)

    data = {
        "org_acronym": org_acronym,
        "page": page,
        "per_page": per_page,
        "since_timestamp": since_timestamp,
    }

    try:
        response = requests.post(
            "https://api.staging-cd.crossfeed.cyber.dhs.gov/pe/apiv1/get_mdl_was_findings",
            headers=headers,
            json=data,
            timeout=20,  # Timeout in seconds
        )
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        LOGGER.error("Error fetching DMZ task: %s", e)
        return None


def fetch_dmz_was_finding_data(task_id):
    """Fetch DMZ WAS Finding data for a task."""
    url = "https://api.staging-cd.crossfeed.cyber.dhs.gov/pe/apiv1/get_mdl_was_findings/task/{t_id}".format(
        t_id=task_id
    )

    try:
        response = requests.get(url, headers=headers, timeout=20)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        LOGGER.error("Error fetching DMZ Was Finding data: %s", e)
        return None


def convert_timestamp_to_date(timestamp: str) -> Optional[str]:
    """Convert an ISO 8601 timestamp to a date string in YYYY-MM-DD format."""
    try:
        date_object = datetime.datetime.strptime(timestamp, "%Y-%m-%dT%H:%M:%S")
        formatted_date = date_object.strftime("%Y-%m-%d")
    except Exception:
        formatted_date = None
    return formatted_date


def save_findings_to_db(was_finding_array, org):
    """Save WAS finding data to the mini datalake using Django ORM."""
    if was_finding_array:
        for asset in was_finding_array:
            try:
                (
                    was_finding_object,
                    created,
                ) = WasFindings.objects.update_or_create(
                    finding_uid=asset.get("finding_uid"),
                    defaults={
                        "finding_type": asset.get("finding_type"),
                        "webapp_id": asset.get("webapp_id"),
                        "webapp_url": asset.get("webapp_url"),
                        "webapp_name": asset.get("webapp_name"),
                        "was_org_id": asset.get("was_org_id"),
                        "name": asset.get("name"),
                        "owasp_category": asset.get("owasp_category"),
                        "severity": asset.get("severity"),
                        "times_detected": asset.get("times_detected"),
                        "cvss_v3_attack_vector": asset.get("cvss_v3_attack_vector"),
                        "base_score": asset.get("base_score"),
                        "temporal_score": asset.get("temporal_score"),
                        "fstatus": asset.get("fstatus"),
                        "last_detected": convert_timestamp_to_date(
                            asset.get("last_detected")
                        ),
                        "first_detected": convert_timestamp_to_date(
                            asset.get("first_detected")
                        ),
                        "potential": asset.get("potential"),
                        "cwe_list": asset.get("cwe_list"),
                        "wasc_list": asset.get("wasc_list"),
                        "last_tested": convert_timestamp_to_date(
                            asset.get("last_tested")
                        ),
                        "fixed_date": convert_timestamp_to_date(
                            asset.get("fixed_date")
                        ),
                        "is_ignored": asset.get("is_ignored"),
                        "is_remediated": asset.get("is_remediated"),
                        "url": asset.get("url"),
                        "qid": asset.get("qid"),
                        "response": asset.get("response"),
                    },
                )

            except Exception as e:
                LOGGER.error("Error saving Was Finding: %s", e)
